src/Runner.Worker/ActionManager.cs

```diff
@@ -596,73 +597,89 @@ private async Task<bool> RepoExistsAsync(IExecutionContext executionContext, Pip
-        private async Task<Dictionary<string, ActionDownloadInfo>> GetDownloadInfoAsync(IExecutionContext executionContext, List<Pipelines.ActionStep> actions)
+        private async Task<IDictionary<string, WebApi.ActionDownloadInfo>> GetDownloadInfoAsync(IExecutionContext executionContext, List<Pipelines.ActionStep> actions)
         {
-            var result = new Dictionary<string, ActionDownloadInfo>(StringComparer.OrdinalIgnoreCase);
+            executionContext.Output("Getting action download info");
 
-            var configurationStore = HostContext.GetService<IConfigurationStore>();
-            var runnerSettings = configurationStore.GetSettings();
-            var apiUrl = GetApiUrl(executionContext);
-            var accessToken = executionContext.GetGitHubContext("token");
-            var base64EncodingToken = Convert.ToBase64String(Encoding.UTF8.GetBytes($"x-access-token:{accessToken}"));
-            var authorization = $"Basic {base64EncodingToken}";
+            // Convert to action reference
+            var actionReferences = actions
+                .GroupBy(x => GetDownloadInfoLookupKey(x))
+                .Where(x => !string.IsNullOrEmpty(x.Key))
+                .Select(x =>
+                {
+                    var action = x.First();
+                    var repositoryReference = action.Reference as Pipelines.RepositoryPathReference;
+                    ArgUtil.NotNull(repositoryReference, nameof(repositoryReference));
+                    return new WebApi.ActionReference
+                    {
+                        NameWithOwner = repositoryReference.Name,
+                        Ref = repositoryReference.Ref,
+                    };
+                })
+                .ToList();
 
-            foreach (var action in actions)
+            // Nothing to resolve?
+            if (actionReferences.Count == 0)
             {
-                var lookupKey = GetDownloadInfoLookupKey(action);
-                if (string.IsNullOrEmpty(lookupKey) || result.ContainsKey(lookupKey))
+                return new Dictionary<string, WebApi.ActionDownloadInfo>();
+            }
+
+            // Resolve download info
+            var jobServer = HostContext.GetService<IJobServer>();
+            var actionDownloadInfos = default(WebApi.ActionDownloadInfoCollection);
+            for (var attempt = 1; attempt <= 3; attempt++)
+            {
+                try
                 {
-                    continue;
+                    actionDownloadInfos = await jobServer.ResolveActionDownloadInfoAsync(executionContext.Plan.ScopeIdentifier, executionContext.Plan.PlanType, executionContext.Plan.PlanId, new WebApi.ActionReferenceList { Actions = actionReferences }, executionContext.CancellationToken);
+                    break;
                 }
+                catch (Exception ex) when (attempt < 3)
+                {
+                    executionContext.Output(ex.Message);
```

Comment:
maybe: `executionContext.Output($"Fail to resolve actions download info, error: {ex.Message}");`